File: sdks/typescript/src/core/composite-ops.ts
/**
 * DCP v2.0 Composite Signature Operations.
 *
 * Implements cryptographically-bound hybrid signatures where the PQ signature
 * covers the classical signature, preventing stripping attacks.
 *
 * Binding protocol:
 *   Step 1: classical_sig = Classical.sign(context || 0x00 || payload)
 *   Step 2: pq_sig = PQ.sign(context || 0x00 || payload || classical_sig)
 *
 * Verification:
 *   Step 1: Verify pq_sig over (context || 0x00 || payload || classical_sig)
 *   Step 2: Verify classical_sig over (context || 0x00 || payload)
 */

import type { CryptoProvider } from './crypto-provider.js';
import type { CompositeSignature, SignatureEntry } from './composite-sig.js';
import type { DcpContext } from './domain-separation.js';
import { domainSeparatedMessage } from './domain-separation.js';
import { AlgorithmRegistry } from './crypto-registry.js';

export interface CompositeKeyInfo {
  kid: string;
  secretKeyB64: string;
  publicKeyB64: string;
  alg: string;
}

export interface CompositeKeyPair {
  classical: CompositeKeyInfo;
  pq: CompositeKeyInfo;
}

export interface CompositeVerifyResult {
  valid: boolean;
  classical_valid: boolean;
  pq_valid: boolean;
}

function concat(...arrays: Uint8Array[]): Uint8Array {
  const total = arrays.reduce((sum, a) => sum + a.length, 0);
  const result = new Uint8Array(total);
  let offset = 0;
  for (const a of arrays) {
    result.set(a, offset);
    offset += a.length;
  }
  return result;
}

/**
 * Produce a composite-bound hybrid signature.
 *
 * The PQ signature covers the classical signature, binding them together.
 * Stripping either component causes both to fail verification.
 */
export async function compositeSign(
  registry: AlgorithmRegistry,
  context: DcpContext | string,
  canonicalPayloadBytes: Uint8Array,
  keys: CompositeKeyPair,
): Promise<CompositeSignature> {
  const classical = registry.getSigner(keys.classical.alg);
  const pq = registry.getSigner(keys.pq.alg);

  const dsm = domainSeparatedMessage(context, canonicalPayloadBytes);

  const classicalSigBytes = await classical.sign(dsm, keys.classical.secretKeyB64);

  const compositeMessage = concat(dsm, classicalSigBytes);
  const pqSigBytes = await pq.sign(compositeMessage, keys.pq.secretKeyB64);

  const classicalEntry: SignatureEntry = {
    alg: keys.classical.alg,
    kid: keys.classical.kid,
    sig_b64: Buffer.from(classicalSigBytes).toString('base64'),
  };

  const pqEntry: SignatureEntry = {
    alg: keys.pq.alg,
    kid: keys.pq.kid,
    sig_b64: Buffer.from(pqSigBytes).toString('base64'),
  };

  return {
    classical: classicalEntry,
    pq: pqEntry,
    binding: 'pq_over_classical',
  };
}

/**
 * Produce a classical-only composite signature (transition mode).
 */
export async function classicalOnlySign(
  registry: AlgorithmRegistry,
  context: DcpContext | string,
  canonicalPayloadBytes: Uint8Array,
  key: CompositeKeyInfo,
): Promise<CompositeSignature> {
  const provider = registry.getSigner(key.alg);
  const dsm = domainSeparatedMessage(context, canonicalPayloadBytes);
  const sigBytes = await provider.sign(dsm, key.secretKeyB64);

  return {
    classical: {
      alg: key.alg,
      kid: key.kid,
      sig_b64: Buffer.from(sigBytes).toString('base64'),
    },
    pq: null,
    binding: 'classical_only',
  };
}

export type VerifyStrategy = 'parallel' | 'pq_first';

/**
 * Verify a composite-bound hybrid signature.
 *
 * For `pq_over_classical` binding:
 *   1. Verify PQ sig over (dsm || classical_sig) — fails if classical was tampered
 *   2. Verify classical sig over dsm
 *
 * Strategies:
 *   - `parallel` (default): Both verified concurrently via Promise.all
 *   - `pq_first`: PQ verified first; if it fails, skip classical (fast-fail)
 */
export async function compositeVerify(
  registry: AlgorithmRegistry,
  context: DcpContext | string,
  canonicalPayloadBytes: Uint8Array,
  compositeSig: CompositeSignature,
  classicalPubkeyB64: string,
  pqPubkeyB64?: string,
  strategy: VerifyStrategy = 'parallel',
): Promise<CompositeVerifyResult> {
  const dsm = domainSeparatedMessage(context, canonicalPayloadBytes);

  if (compositeSig.binding === 'classical_only') {
    if (compositeSig.pq !== null) {
      return { valid: false, classical_valid: false, pq_valid: false };
    }
    const classicalProvider = registry.getSigner(compositeSig.classical.alg);
    const classicalSigBytes = Buffer.from(compositeSig.classical.sig_b64, 'base64');
    const classicalValid = await classicalProvider.verify(
      dsm,
      new Uint8Array(classicalSigBytes),
      classicalPubkeyB64,
    );
    return { valid: classicalValid, classical_valid: classicalValid, pq_valid: false };
  }

  if (compositeSig.binding !== 'pq_over_classical') {
    return { valid: false, classical_valid: false, pq_valid: false };
  }

  if (!compositeSig.pq || !pqPubkeyB64) {
    return { valid: false, classical_valid: false, pq_valid: false };
  }

  const classicalProvider = registry.getSigner(compositeSig.classical.alg);
  const pqProvider = registry.getSigner(compositeSig.pq.alg);

  const classicalSigBytes = new Uint8Array(
    Buffer.from(compositeSig.classical.sig_b64, 'base64'),
  );
  const pqSigBytes = new Uint8Array(
    Buffer.from(compositeSig.pq.sig_b64, 'base64'),
  );

  const compositeMessage = concat(dsm, classicalSigBytes);

  if (strategy === 'pq_first') {
    const pqValid = await pqProvider.verify(compositeMessage, pqSigBytes, pqPubkeyB64);
    if (!pqValid) {
      return { valid: false, classical_valid: false, pq_valid: false };
    }
    const classicalValid = await classicalProvider.verify(dsm, classicalSigBytes, classicalPubkeyB64);
    return {
      valid: classicalValid && pqValid,
      classical_valid: classicalValid,
      pq_valid: pqValid,
    };
  }

  const [classicalValid, pqValid] = await Promise.all([
    classicalProvider.verify(dsm, classicalSigBytes, classicalPubkeyB64),
    pqProvider.verify(compositeMessage, pqSigBytes, pqPubkeyB64),
  ]);

  return {
    valid: classicalValid && pqValid,
    classical_valid: classicalValid,
    pq_valid: pqValid,
  };
}

File: sdks/python/dcp_ai/v2/composite_ops.py
"""
DCP v2.0 Composite Signature Operations.

Implements cryptographically-bound hybrid signatures where the PQ signature
covers the classical signature, preventing stripping attacks.
"""

from __future__ import annotations

import base64
from typing import NamedTuple

from dcp_ai.v2.composite_sig import CompositeSignature, SignatureEntry
from dcp_ai.v2.crypto_provider import CryptoProvider
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import domain_separated_message


class CompositeKeyInfo(NamedTuple):
    kid: str
    alg: str
    secret_key_b64: str
    public_key_b64: str


class CompositeVerifyResult(NamedTuple):
    valid: bool
    classical_valid: bool
    pq_valid: bool


async def composite_sign(
    registry: AlgorithmRegistry,
    context: str,
    canonical_payload_bytes: bytes,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
) -> CompositeSignature:
    """Produce a composite-bound hybrid signature.

    The PQ signature covers the classical signature, binding them together.
    """
    classical_provider = registry.get_signer(classical_key.alg)
    pq_provider = registry.get_signer(pq_key.alg)

    dsm = domain_separated_message(context, canonical_payload_bytes)

    classical_sig = await classical_provider.sign(dsm, classical_key.secret_key_b64)

    composite_message = dsm + classical_sig
    pq_sig = await pq_provider.sign(composite_message, pq_key.secret_key_b64)

    return CompositeSignature(
        classical=SignatureEntry(
            alg=classical_key.alg,
            kid=classical_key.kid,
            sig_b64=base64.b64encode(classical_sig).decode("ascii"),
        ),
        pq=SignatureEntry(
            alg=pq_key.alg,
            kid=pq_key.kid,
            sig_b64=base64.b64encode(pq_sig).decode("ascii"),
        ),
        binding="pq_over_classical",
    )


async def classical_only_sign(
    registry: AlgorithmRegistry,
    context: str,
    canonical_payload_bytes: bytes,
    key: CompositeKeyInfo,
) -> CompositeSignature:
    """Produce a classical-only composite signature (transition mode)."""
    provider = registry.get_signer(key.alg)
    dsm = domain_separated_message(context, canonical_payload_bytes)
    sig = await provider.sign(dsm, key.secret_key_b64)

    return CompositeSignature(
        classical=SignatureEntry(
            alg=key.alg,
            kid=key.kid,
            sig_b64=base64.b64encode(sig).decode("ascii"),
        ),
        pq=None,
        binding="classical_only",
    )


async def composite_verify(
    registry: AlgorithmRegistry,
    context: str,
    canonical_payload_bytes: bytes,
    composite_sig: CompositeSignature,
    classical_pubkey_b64: str,
    pq_pubkey_b64: str | None = None,
    strategy: str = "parallel",
) -> CompositeVerifyResult:
    """Verify a composite-bound hybrid signature.

    For pq_over_classical binding, both signatures are verified and the PQ
    signature's coverage of the classical signature is confirmed.

    Strategies:
      - "parallel" (default): Both verified concurrently via asyncio.gather
      - "pq_first": PQ verified first; if it fails, skip classical (fast-fail)
    """
    import asyncio

    dsm = domain_separated_message(context, canonical_payload_bytes)

    if composite_sig["binding"] == "classical_only":
        if composite_sig["pq"] is not None:
            return CompositeVerifyResult(valid=False, classical_valid=False, pq_valid=False)
        classical_provider = registry.get_signer(composite_sig["classical"]["alg"])
        classical_sig_bytes = base64.b64decode(composite_sig["classical"]["sig_b64"])
        classical_valid = await classical_provider.verify(
            dsm, classical_sig_bytes, classical_pubkey_b64
        )
        return CompositeVerifyResult(
            valid=classical_valid, classical_valid=classical_valid, pq_valid=False
        )

    if composite_sig["binding"] != "pq_over_classical":
        return CompositeVerifyResult(valid=False, classical_valid=False, pq_valid=False)

    if composite_sig["pq"] is None or pq_pubkey_b64 is None:
        return CompositeVerifyResult(valid=False, classical_valid=False, pq_valid=False)

    classical_provider = registry.get_signer(composite_sig["classical"]["alg"])
    pq_provider = registry.get_signer(composite_sig["pq"]["alg"])

    classical_sig_bytes = base64.b64decode(composite_sig["classical"]["sig_b64"])
    pq_sig_bytes = base64.b64decode(composite_sig["pq"]["sig_b64"])

    composite_message = dsm + classical_sig_bytes

    if strategy == "pq_first":
        pq_valid = await pq_provider.verify(
            composite_message, pq_sig_bytes, pq_pubkey_b64
        )
        if not pq_valid:
            return CompositeVerifyResult(valid=False, classical_valid=False, pq_valid=False)
        classical_valid = await classical_provider.verify(
            dsm, classical_sig_bytes, classical_pubkey_b64
        )
        return CompositeVerifyResult(
            valid=classical_valid and pq_valid,
            classical_valid=classical_valid,
            pq_valid=pq_valid,
        )

    classical_valid, pq_valid = await asyncio.gather(
        classical_provider.verify(dsm, classical_sig_bytes, classical_pubkey_b64),
        pq_provider.verify(composite_message, pq_sig_bytes, pq_pubkey_b64),
    )

    return CompositeVerifyResult(
        valid=classical_valid and pq_valid,
        classical_valid=classical_valid,
        pq_valid=pq_valid,
    )
